src/esi/Esi.cc

```diff
@@ -1256,8 +1259,24 @@ ESIContext::parse()
 
         PROF_start(esiParsing);
 
-        while (buffered.getRaw() && !flags.error)
-            parseOneBuffer();
+        try {
+            while (buffered.getRaw() && !flags.error)
+                parseOneBuffer();
+
+        } catch (Esi::ErrorDetail &errMsg) { // FIXME: non-const for c_str()
+            // level-2: these are protocol/syntax errors from upstream
+            debugs(86, 2, "WARNING: ESI syntax error: " << errMsg);
+            setError();
+            setErrorMessage(errMsg.c_str());
+
+        } catch (...) {
+            // DBG_IMPORTANT because these are local issues the admin needs to fix
+            static FadingCounter logEntries(); // TODO: set horizon less than infinity
```

Comment:
This overly specific construction probably causes this build [error](https://semaphoreci.com/squid/squid/branches/pull-request-411/builds/18):

```
src/esi/Esi.cc:1275:28: error: request for member ‘count’ in ‘logEntries’, which is of non-class type ‘FadingCounter()’
             if (logEntries.count(1) < 100)
```

I suggest to KISS:

```suggestion
            static FadingCounter logEntries; // TODO: set horizon less than infinity
```

The whole FadingCounter API should be adjusted if we are going to use it for limiting debugs() statements, but those adjustments are outside this PR scope. If you prefer to replace this FadingCounter with a basic `unsigned int` or `size_t` counter, please feel free to do so without triggering another review!

Reply:
Agreed on the need for adjustments to the fader class. The primary one is as noted in the TODO.

FTR; my reasoning is this:
 If we use a simple counter and stop reporting past a threshold we could hit problems with one error happening N times filling the counter, then silence some other errors. That can still happen with the fader but self-corrects as the fading happens, allowing the suppressed problems to be reported at some point.
 Making a more performant fader (if we can) would be good not just in debugs() case but for the existing users of the FadingCounter class. So TODO on a redesign is better than a second implementation or a bad counter.